multinet/uploaders/d3_json.py

```diff
@@ -0,0 +1,99 @@
+"""Multinet uploader for nested JSON files."""
+import json
+from io import StringIO
+import pandas as pd
+from collections import OrderedDict
+
+from .. import db, util
+from ..errors import ValidationFailed
+from ..util import decode_data
+
+from flask import Blueprint, request
+from flask import current_app as app
+
+# Import types
+from typing import Any
+
+bp = Blueprint("d3_json", __name__)
+bp.before_request(util.require_db)
+
+
+def validate_d3_json(data):
+    """Perform any necessary d3 json validation, and return appropriate errors."""
+    data_errors = []
+
+    # Check the structure of the uploaded file is what we expect
+    if "nodes" not in data.keys() or "links" not in data.keys():
+        data_errors.append({"error": "structure"})
+
+    # Check that links are in source -> target form
+    if not all(
+        "source" in row.keys() and "target" in row.keys() for row in data["links"]
+    ):
+        data_errors.append({"error": "invalid_link_keys"})
+
+    # Check that the keys for each dictionary match
+    if not all(data["links"][0].keys() == row.keys() for row in data["links"]):
+        data_errors.append({"error": "inconsistent_link_keys"})
+
+    # Check for duplicated nodes
+    ids = [row["id"] for row in data["nodes"]]
+    if len(ids) != len(set(ids)):
+        data_errors.append({"error": "node_duplicates"})
+
+    # Duplicated edges are okay, they might encode different relationships, but
+    # that should be specified in the attributes. Thus duplicated dicts are not okay
+    if len(data["links"]) != len(set([tuple(x.items()) for x in data["links"]])):
+        data_errors.append({"error": "link_duplicates"})
+
+    if len(data_errors) > 0:
+        raise ValidationFailed(data_errors)
```

Comment:
I think instead of raising the exception here, it's better to return the list of errors. Then, at the point where validation is needed, the caller can look at the result and decide what to do with it (in this case, simply raise `ValidationFailed()` as you do here).

This has two major benefits:
1. It makes the usage of this function more explicit where it's used, rather than making the reader of the code guess or have to look up that it raises an exception.
2. It makes testing this function simpler, as you don't have to catch exceptions to validate this function.

If you're copying the validation pattern from the other uploader modules, then I think you should still make this change here as suggested, and file an issue to fix the pattern in the other validators.

Reply:
I agree with all of your points here. I've updated the code for this uploader and have opened a new issue to update the other uploaders #200.